### PR TITLE
[AIRFLOW-6373] Make airflow/utils pylint compatible

### PR DESCRIPTION
**Pull request by turbaszek**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6367

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
This PR make `airflow/utils` pylint compatible.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: scripts/ci/pylint_todo.txt [hunk 1]
```diff
@@ -198,34 +198,7 @@
...
-./airflow/utils/operator_helpers.py
-./airflow/utils/operator_resources.py
-./airflow/utils/sqlalchemy.py
-./airflow/utils/state.py
-./airflow/utils/tests.py
-./airflow/utils/timeout.py
-./airflow/utils/timezone.py
-./airflow/utils/trigger_rule.py
-./airflow/utils/weight_rule.py
+./airflow/utils/log/colored_log.py
```
Comment (turbaszek) [inline scripts/ci/pylint_todo.txt]: I have no idea why but we have cyclic import here:
```
************* Module airflow.utils.log.colored_log
airflow/utils/log/colored_log.py:1:0: R0401: Cyclic import (airflow.models -> airflow.models.baseoperator -> airflow.utils.db) (cyclic-import)
airflow/utils/log/colored_log.py:1:0: R0401: Cyclic import (airflow.models.baseoperator -> airflow.utils.helpers) (cyclic-import)
airflow/utils/log/colored_log.py:1:0: R0401: Cyclic import (airflow.models.baseoperator -> airflow.utils.db -> airflow.models.serialized_dag -> airflow.serialization.serialized_objects) (cyclic-import)
airflow/utils/log/colored_log.py:1:0: R0401: Cyclic import (airflow.models.serialized_dag -> airflow.utils.db) (cyclic-import)
airflow/utils/log/colored_log.py:1:0: R0401: Cyclic import (airflow.models -> airflow.models.taskreschedule -> airflow.utils.db) (cyclic-import)
```

**Comment by turbaszek** [comment]
Or even more cyclic imports?
```
************* Module airflow.cli.commands.config_command
1652airflow/cli/commands/config_command.py:1:0: R0401: Cyclic import (airflow.models -> airflow.models.taskreschedule -> airflow.utils.db) (cyclic-import)
1653airflow/cli/commands/config_command.py:1:0: R0401: Cyclic import (airflow.models.baseoperator -> airflow.utils.db -> airflow.models.serialized_dag -> airflow.serialization.serialized_objects) (cyclic-import)
1654airflow/cli/commands/config_command.py:1:0: R0401: Cyclic import (airflow.models.baseoperator -> airflow.utils.helpers) (cyclic-import)
1655airflow/cli/commands/config_command.py:1:0: R0401: Cyclic import (airflow.models -> airflow.models.baseoperator -> airflow.utils.db) (cyclic-import)
1656airflow/cli/commands/config_command.py:1:0: R0401: Cyclic import (airflow.models -> airflow.models.taskreschedule -> airflow.utils.db -> airflow.models.serialized_dag -> airflow.serialization.serialized_objects) (cyclic-import)
1657airflow/cli/commands/config_command.py:1:0: R0401: Cyclic import (airflow.models.serialized_dag -> airflow.utils.db) (cyclic-import)
```

**Comment by potiuk** [comment]
The reasons are .... cyclic imports :)

The problem is that sometimes we have too many grouped classes in one module where it is really not supposed to be groupped. For example the baseoperator->helpers cyclic import is caused by "chain" method in helpers. BaseOperator uses helpers module and helpers.chain() method uses BaseOperator. This all can be resolved by careful untangling of modules. Sometimes what helps is putting together things that are actually coupled by historically are in separate modules. I have some experience with that and will try to add a few commits on top of yours to solve it.

**Comment by potiuk** [comment]
In fact the baseoperator -> helpers root cause is that baseoperator uses helpers.validate_key as the only method from helpers. There are other methods in helpers that use BaseOperator and it makes sense to import BaseOperator there at top level and move validate_key somewhere else

**Comment by potiuk** [comment]
@nuclearpinguin See ^^ here I solved the baseoperator -> helpers import. That actually will enable to add a few more BaseOperators (typehints) in helpers and move BaseOperator import to top level import (next commit that follows)

**Comment by potiuk** [comment]
@nuclearpinguin ^^ Here you can see  how it helped to add type annotation in the BaseOperator-related functions - now BaseOperator can be top-level and can be used in method signature.

The next thing will be to move those two BaseOperator methods to BaseOperator class. They actually belong there  - helpers are methods that can be imported in many places and should be fairly "standalone" so actually I think making helper dependent on BaseOperator is bad idea.

**Comment by potiuk** [comment]
And now even better ^^ the two methods from helpers which are tightly coupled with BaseOperator are now static methods in BaseOperator itself.

**Comment by turbaszek** [comment]
There are still cyclic imports and I start to doubt in pylint messages, for example:
```
airflow/utils/log/es_task_handler.py:1:0: R0401: Cyclic import (airflow.models -> airflow.models.taskreschedule -> airflow.utils.db) (cyclic-import)
```
and I don't see how it is possible as `airflow.utils.db` do not import anything from `airflow.models` unless you count the import in functions... more info
https://github.com/PyCQA/pylint/issues/850

**Comment by potiuk** [comment]
Yep. I have not finished yet. In the last one I brought back the key_validation back to helpers (now we can do it as the BaseOperator is not used there).

**Comment by potiuk** [comment]
I think pylint is VERY right about those cyclic imports.

For me the problem is the db module itself.. It's typical manifestation of cyclical imports where you thing various methods are coupled and put them together for a wrong reason. 
If you look what db module now is: "let's use it for any kind of db  operations". 

But in fact there are two different purpose it serves now:
* provide methods that are used everywhere for DB operations (including models) - `provide_session` is such method. Those methods are USED by the models.
* provide methods that are used to group together db methods for multiple models (those methods are USING models).

Putting those two types of methods introduces coupling between them unnecessarily. That's exactly what is causing the cyclic imports. 

We should split the db module into two independent modules. I am going to do just that shortly.

**Comment by potiuk** [comment]
And merge_conn method should be moved to Connection.

**Comment by turbaszek** [comment]
I just removed inner imports in `airflow.db` result: no cyclic imports.  There is a difference between real cyclic import which will be reported by Python and those "probably a cyclic import" raised by pylint.

However, I agree that it's a good idea to remove session stuff to some `session` module.

**Comment by potiuk** [comment]
That's the point. Those ARE real (not probable) cyclic imports. And of course Pylint reports more/different kinds cyclic imports than python parser - otherwise we would not even get there at the first place because pylint would fail parsing the code. Pylint deliberately reports the cycles that developers have hidden by using local imports. But they are only hidden - they are logical cycles. 

They are still there but in order to stop python parser from failing at import time developer deferred resolving imports to later phase when method is executed. But real coupling between the MODULE containing the method and imported class remains. It's the logical cycle that is there: resetdb() (from db)  method uses 'models' which are using provide_session() (again from db). This is a real cycle. And it does not need to be ther

**Comment by turbaszek** [comment]
After a discussion I extracted session related stuff to a new `session` module #6938 . I will rebase after the merge.

Thanks @potiuk !

**Comment by ashb** [comment]
Can you try to keep PRs to a single thing please? This is at least three? Or five! prs - white space changes, lots of doc comments, unrelated pylint changes, touching a file that was otherwise not changed etc.

**Comment by ashb** [comment]
Wait did this get renamed? When I first opened it it was about just moving something around I thought

**Comment by turbaszek** [comment]
After merging #6938 I propose to include in this PR only the first commit. I agree with @ashb that all additional changes should be done in separate one, before or after this PR.

**Comment by potiuk** [comment]
Yep. I will move my commits as next step after #6938 . I will do it shortly.

File: airflow/utils/tests.py [hunk 2]
```diff
@@ -28,19 +27,11 @@
 from airflow.utils.decorators import apply_defaults
 
 
-def skipUnlessImported(module, obj):
```
Comment (turbaszek) [inline airflow/utils/tests.py]: Was never used in project.

File: airflow/utils/configuration.py [hunk 3]
```diff
@@ -24,7 +24,11 @@
 from airflow.configuration import conf
 
 
-def tmp_configuration_copy(chmod=0o600, include_env=True, include_cmds=True):
+def tmp_configuration_copy(  # pylint: disable=unused-argument
+    chmod=0o600,
+    include_env=True,
```
Comment (mik-laj) [inline airflow/utils/configuration.py]: Should we remove unused argument?

**Comment by potiuk** [comment]
Unit tests failing  :(

File: airflow/task/task_runner/base_task_runner.py [hunk 4]
```diff
@@ -63,9 +63,7 @@ def __init__(self, local_task_job):
             # want to have to specify them in the sudo call - they would show
             # up in `ps` that way! And run commands now, as the other user
             # might not be able to run the cmds to get credentials
-            cfg_path = tmp_configuration_copy(chmod=0o600,
-                                              include_env=True,
-                                              include_cmds=True)
```
Comment (turbaszek) [inline airflow/task/task_runner/base_task_runner.py:68]: Those args were unused.

**Comment by potiuk** [comment]
Still a few failures... But we are getting there!

**Review by potiuk — APPROVED** [review]

**Pull request merged**